### PR TITLE
Learner Integration

### PR DESCRIPTION
**Pull request by gabbard**

### CODE REVIEW

File: adam/learner/attributes.py [hunk 1]
```diff
@@ -11,16 +13,89 @@
...
+            language_perception_semantic_alignment.language_concept_alignment
+        )
+        # Find all objects we have recognized...
+        for (
+            object_node,
+            span_for_object,
+        ) in language_concept_alignment.node_to_language_span.items():
+            if isinstance(object_node, ObjectSemanticNode):
+                # Any words immediately before them or after them are candidate attributes.
+                # TODO: make issue to track this somewhat English-specific assumption
```
Comment (LichMaster98) [inline adam/learner/attributes.py:61]: Reminder to turn this into an issue. I'm going to mark future ones of these with a 🗒️ emote

File: adam/learner/integrated_learner.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        for sub_learner in [
+            self.object_learner,
+            self.attribute_learner,
+            self.relation_learner,
+        ]:
+            if sub_learner:
+                # Currently we do not attempt to learn static things from dynamic situations
+                # because the static learners do not know how to deal with the temporal
+                # perception graph edge wrappers.
+                # @Reviewers: do not let me merge without making an issue for this.
```
Comment (LichMaster98) [inline adam/learner/integrated_learner.py:131]: 🗒️

File: adam/learner/objects.py [hunk 3]
```diff
@@ -11,31 +11,93 @@
...
+                perception_semantic_alignment.perception_graph
+            ),
+            new_nodes=[],
+        )
+
+    def _candidate_templates(
+        self, language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment
+    ) -> AbstractSet[BoundSurfaceTemplate]:
+        # TODO: make an issue to track that we can only learn single words for objects
+        # at the moment
```
Comment (LichMaster98) [inline adam/learner/objects.py:84]: 🗒️

File: adam/learner/__init__.py [hunk 4]
```diff
@@ -167,3 +172,27 @@ def graph_without_learner(perception_graph: PerceptionGraph) -> PerceptionGraph:
         islands = list(isolates(graph))
         graph.remove_nodes_from(islands)
     return PerceptionGraph(graph)
+
+
+class NewStyleLearner(ABC):
```
Comment (gabbard) [inline adam/learner/__init__.py:177]: @spigo900 : Please add docstrings for these methods.

File: adam/learner/alignments.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+                aligned_token_span.start
+            ] = object_node_to_template_variable[matched_object_node]
+
+        return SurfaceTemplate(
+            template_elements, determiner_prefix_slots=determiner_prefix_slots
+        )
+
+
+@attrs(frozen=True)
+class PerceptionSemanticAlignment:
```
Comment (gabbard) [inline adam/learner/alignments.py:270]: @spigo900 : Please add a docstring (henceforth 📖 )

File: adam/learner/alignments.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+            if isinstance(node, SemanticNode):
+                if not node in self.semantic_nodes:
+                    raise RuntimeError(
+                        "All semantic nodes appearing in the perception graph must "
+                        "also be in semantic_nodes"
+                    )
+
+
+@attrs(frozen=True, kw_only=True)
+class LanguagePerceptionSemanticAlignment:
```
Comment (gabbard) [inline adam/learner/alignments.py:304]: 📖

File: adam/learner/attributes.py (hunk 1 above)
Comment (gabbard) [reply adam/learner/attributes.py:61]: @spigo900 : Please make the requested issue and add a pointer to it in the code

File: adam/learner/integrated_learner.py (hunk 2 above)
Comment (gabbard) [reply adam/learner/integrated_learner.py:132]: @spigo900 : Please create and reference the issue.

File: adam/learner/integrated_learner.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        attributes_we_can_express = (
+            [
+                attribute
+                for attribute in learner_semantics.objects_to_attributes[object_node]
+                if self.attribute_learner.templates_for_concept(attribute.concept)
+            ]
+            if self.attribute_learner
+            else []
+        )
+        # TODO: deal with relations
```
Comment (gabbard) [inline adam/learner/integrated_learner.py:231]: @spigo900 : Please make and reference an issue about how we can express "top-level" relations, but we can't yet express relations as modifiers of objects embedded in action descriptions, etc.

File: adam/ontology/ontology.py [hunk 8]
```diff
@@ -152,8 +158,36 @@ def nodes_with_properties(
             if self.has_all_properties(
                 node, required_properties, banned_properties=banned_properties
             )
+            and node not in banned_ontology_types
+            and not self.descends_from(node, banned_ontology_types)
         )
 
+    def descends_from(
```
Comment (gabbard) [inline adam/ontology/ontology.py:165]: @spigo900 : Please add a unit test for this.

File: adam/ontology/ontology.py [hunk 9]
```diff
@@ -122,22 +122,28 @@ def nodes_with_properties(
         required_properties: Iterable["OntologyNode"] = immutableset(),
         *,
         banned_properties: AbstractSet["OntologyNode"] = immutableset(),
+        banned_ontology_types: AbstractSet["OntologyNode"] = immutableset(),
```
Comment (gabbard) [inline adam/ontology/ontology.py:125]: @spigo900 : Please add a unit test for this.

File: adam/learner/attributes.py (hunk 1 above)
Comment (spigo900) [reply adam/learner/attributes.py]: @gabbard Created the issue: #791.

File: adam/learner/integrated_learner.py (hunk 2 above)
Comment (spigo900) [reply adam/learner/integrated_learner.py]: @gabbard Done. It's issue #792.

File: adam/learner/objects.py (hunk 3 above)
Comment (spigo900) [reply adam/learner/objects.py]: @gabbard Issue #793.

File: adam/learner/__init__.py (hunk 4 above)
Comment (spigo900) [reply adam/learner/__init__.py:177]: @gabbard Done. Should I also add a docstring for the class?

File: adam/learner/alignments.py (hunk 5 above)
Comment (spigo900) [reply adam/learner/alignments.py:270]: @gabbard Done.

File: adam/learner/alignments.py (hunk 6 above)
Comment (spigo900) [reply adam/learner/alignments.py:311]: @gabbard Done.

Also, I noticed that the docstring on `LanguageConceptAlignment` seems to be [out of date][old_docstring]. It's written as if the class represented a direct alignment between language and perception whereas now it's between language and concepts. Should I update that as well?

[old_docstring]: https://github.com/isi-vista/adam/blob/91d5bb7bb9c2e0a0e8679e52618f2468be76d391/adam/learner/alignments.py#L35

File: adam/learner/integrated_learner.py (hunk 7 above)
Comment (spigo900) [reply adam/learner/integrated_learner.py]: @gabbard Issue #794.

File: adam/ontology/ontology.py (hunk 8 above)
Comment (spigo900) [reply adam/ontology/ontology.py:165]: @gabbard Done, [commit here](https://github.com/isi-vista/adam/pull/773/commits/91d5bb7bb9c2e0a0e8679e52618f2468be76d391). Does it make sense to use the phase 1 ontology in these tests, or should I rewrite them to use an artificial ontology?

File: adam/ontology/ontology.py (hunk 9 above)
Comment (spigo900) [reply adam/ontology/ontology.py:125]: Done.

File: adam/learner/__init__.py (hunk 4 above)
Comment (gabbard) [reply adam/learner/__init__.py:177]: Not yet - it's probably going to get collapsed with another one before this merges.

File: adam/learner/alignments.py (hunk 6 above)
Comment (gabbard) [reply adam/learner/alignments.py:311]: Please do

File: adam/ontology/ontology.py (hunk 8 above)
Comment (gabbard) [reply adam/ontology/ontology.py:165]: You can use Phase 1

File: adam/learner/alignments.py (hunk 6 above)
Comment (spigo900) [reply adam/learner/alignments.py:312]: @gabbard [Done](https://github.com/isi-vista/adam/blob/787375cb1a21b3a0c1a3d92ed456ee22802dbc9a/adam/learner/alignments.py#L35).

**Pull request merged**